### PR TITLE
1822 impor observed data crash fixed

### LINKED ISSUE #1822: Importing observed data crash
@PavelBal commented on [Wed Nov 09 2022](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/2417)

---

@PavelBal commented on [Wed Nov 09 2022](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/2417#issuecomment-1308888686)

[pksim-2417-importerCrash.zip](https://github.com/Open-Systems-Pharmacology/PK-Sim/files/9972230/pksim-2417-importerCrash.zip)

Trying to import with the provided excel file and configuration (accept that some columns cannot be mapped):

Application:
PK-Sim® 11.1.78

Object reference not set to an instance of an object.

Stack trace:
```
at OSPSuite.Infrastructure.Import.Core.DataFormat.AbstractColumnsDataFormat.parseMappingOnSameGivenColumn(MappingDataFormatParameter currentParameter, DataSheet dataSheet, UnformattedRow row)
   at OSPSuite.Infrastructure.Import.Core.DataFormat.AbstractColumnsDataFormat.<>c__DisplayClass27_2.<parseMappings>b__1(UnformattedRow row)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at OSPSuite.Infrastructure.Import.Core.DataFormat.AbstractColumnsDataFormat.parseMappings(IEnumerable`1 rawDataSet, DataSheet dataSheet, Cache`2 columnInfos)
   at OSPSuite.Infrastructure.Import.Core.DataFormat.AbstractColumnsDataFormat.<>c__DisplayClass26_0.<buildDataSets>b__0(List`1 rows)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at OSPSuite.Infrastructure.Import.Core.DataSet.AddData(IEnumerable`1 range)
   at OSPSuite.Infrastructure.Import.Core.DataSheetCollection.GetDataSets(IDataFormat format, ColumnInfoCache columnInfos)
   at OSPSuite.Infrastructure.Import.Services.Importer.AddFromFile(IDataFormat format, DataSheetCollection dataSheets, ColumnInfoCache columnInfos, IDataSource alreadyExisting)
   at OSPSuite.Infrastructure.Import.Core.DataSource.AddSheets(DataSheetCollection dataSheets, ColumnInfoCache columnInfos, String filter)
   at OSPSuite.Presentation.Presenters.Importer.ImporterPresenter.loadSheets(IDataSourceFile dataSourceFile, IReadOnlyList`1 sheetNames, String filter, String selectedNamingConvention)
   at OSPSuite.Presentation.Presenters.Importer.ImporterPresenter.loadSheetsFromDataPresenter(Object sender, ImportSheetsEventArgs args)
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at OSPSuite.Presentation.Presenters.Importer.ImporterDataPresenter.ImportDataForConfirmation(String sheetName)
   at OSPSuite.UI.Views.Importer.ImporterDataView.onButtonImportClicked(Object sender, EventArgs e)
   at OSPSuite.UI.Controls.BaseUserControl.<>c__DisplayClass26_0`1.<OnEvent>b__0()
   at OSPSuite.UI.Controls.BaseUserControl.<>c__DisplayClass23_0.<OnEvent>b__0()
   at OSPSuite.Utility.Extensions.ExceptionExtensions.DoWithinExceptionHandler(Object callerObject, Action actionToExecute)
```

---

@georgeDaskalakis commented on [Fri Dec 09 2022](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/2417#issuecomment-1344086664)

@PavelBal the whole issue comes from the lloq of the "Measurement" being mapped to the "LLOQ" excel column in the configuration. In the given excel such a column does not exist. I am fixing this check in code, but just till then a workaround is to manually correct the LLOQ mapping after applying the configuration.

---

@georgeDaskalakis commented on [Fri Dec 09 2022](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/2417#issuecomment-1344162344)

@PavelBal @Yuri05 @msevestre @rwmcintosh regarding the fix of this:

The LLOQ mapping will not be automatically cleared when applying the configuration, but we are going to notify the user about the non existent "LLOQ" excel-column when he is trying to load the sheet - then he should have the chance to clear that mapping himself.

### PR DESCRIPTION
**Pull request by georgeDaskalakis**
closes #1822

### CODE REVIEW

File: src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs [hunk 1]
```diff
@@ -94,7 +93,8 @@ private void setSecondaryColumnUnit(ColumnInfoCache columnInfos)
             foreach (var relatedColumn in columnInfos.RelatedColumnsFrom(column.Name))
             {
                var relatedParameter = mappings.FirstOrDefault(p => p.ColumnName == relatedColumn.Name);
-               if (relatedParameter != null && (relatedParameter.MappedColumn.Unit == null || relatedParameter.MappedColumn.Unit.SelectedUnit == UnitDescription.InvalidUnit))
+               if (relatedParameter != null && (relatedParameter.MappedColumn.Unit == null ||
+                                                relatedParameter.MappedColumn.Unit.SelectedUnit == UnitDescription.InvalidUnit))
```
Comment (Yuri05) [inline src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs:97]: `relatedParameter.MappedColumn` cannot be null?

Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs:97]: Judging from the MappingDataFormatParameter constructor I would say no

In DataFormatParameter.cs:
```
      public MappingDataFormatParameter() 
      {
         MappedColumn = new Column();
      }
```

File: src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs [hunk 2]
```diff
@@ -322,6 +332,15 @@ private SimulationPoint parseMappingOnSameColumn(MappingDataFormatParameter curr
 
       private SimulationPoint parseMappingOnSameGivenColumn(MappingDataFormatParameter currentParameter, DataSheet dataSheet, UnformattedRow row)
       {
+         if (!currentParameter.MappedColumn.LloqColumn.IsNullOrEmpty())
+         {
+            ColumnDescription lloqColumnDescription = null;
+            if (!string.IsNullOrEmpty(currentParameter.MappedColumn.LloqColumn))
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs:338]: I don't understand. Why are checking this if you check it already on line 335?
Can we comment as to WHAT we are doing here also. What is the use case. the function is called `parseMappingOnSameGivenColumn` but we are dealing with LLOQ.

**Review by Yuri05 — APPROVED** [review]

File: src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs (hunk 1 above)
Comment (msevestre) [reply src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs:97]: This line of code is very very hard to read. If we are aleady in there, can we simplify it. Create a small method that expresses what it is. Better, this is a a method of the object itself (replated parameter)

File: src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs (hunk 2 above)
Comment (msevestre) [reply src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs:338]: and alco create a variable for currentParameter.MappedColumn.LloqColumn
This is used 4 times in 5 lines

**Review by msevestre — CHANGES_REQUESTED** [review]

File: src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs (hunk 2 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataFormat/AbstractColumnsDataFormat.cs]: OK, so double check removed, added a comment and refactor a bit to make it make more sense (and also avoid some code duplication)

**Review by msevestre — APPROVED** [review]

**Pull request merged**